src/category_theory/flat_functors.lean

```diff
@@ -0,0 +1,266 @@
+
+/--
+Given a cone `c : cone K` and a map `f : X ⟶ F.obj c.X`, we can construct a cone of structured
+arrows over `X` with `f` as the cone point.
+-/
+@[simps]
+def to_cone {X : D} (f : X ⟶ F.obj c.X) :
+  cone (to_diagram (F.map_cone c) ⋙ map f ⋙ pre _ K F) :=
+{ X := mk f, π := { app := λ j, hom_mk (c.π.app j) rfl,
+                    naturality' := λ j k g, by { ext, dsimp, simp } } }
+
+lemma eq_to_hom_right {A : Type*} [category A] {B : Type*} [category B] {T : Type*} [category T]
+  {L : A ⥤ T} {R : B ⥤ T} (X Y : comma L R) (H : X = Y) :
+  comma_morphism.right (eq_to_hom H) = eq_to_hom (by { cases H, refl }) := by { cases H, refl }
+
+local attribute[simp] eq_to_hom_right
+
+/--
+If a cone `s₁` factors through another cone `s₂`, then the two constructed diagrams are actually
+the same.
+-/
+lemma to_diagram_comp_map (s₁ s₂ : cone K)
+  (f : s₁.X ⟶ s₂.X) (H : ∀ (j : J), f ≫ s₂.π.app j = s₁.π.app j) :
+    to_diagram s₂ ⋙ structured_arrow.map f = to_diagram s₁ := by { apply functor.ext, tidy, }
+
+end structured_arrow_cone
+
+section representably_flat
+variables {C : Type u₁} [category.{v₁} C] {D : Type u₂} [category.{v₂} D]
+
+/--
+A functor `F : C ⥤ D` is representably-flat functor if the comma category `(X/F)`
+is cofiltered for each `X : C`.
+-/
+class representably_flat (F : C ⥤ D) : Prop :=
+(cofiltered : ∀ (X : D), is_cofiltered (structured_arrow X F))
+
+instance functor.flat_cofiltered (F : C ⥤ D) [representably_flat F] (X : D) :
+ is_cofiltered (structured_arrow X F) := representably_flat.cofiltered X
+
+end representably_flat
+
+section has_limit
+variables {C : Type u₁} [category.{v₁} C] {D : Type u₂} [category.{v₁} D]
+
+@[priority 100]
+instance cofiltered_of_has_finite_limit [has_finite_limits C] : is_cofiltered C :=
+{ cocone_objs := λ A B, ⟨limits.prod A B, limits.prod.fst, limits.prod.snd, trivial⟩,
+  cocone_maps :=  λ A B f g, ⟨equalizer f g, equalizer.ι f g, equalizer.condition f g⟩,
+  nonempty := ⟨⊤_ C⟩ }
+
+lemma flat_of_preserves_finite_limit [has_limits C] (F : C ⥤ D)
+  (H : ∀ (J : Type v₁) [h : small_category J] [h' : @fin_category J h],
+    @preserves_limits_of_shape _ _ _ _ J h F) : representably_flat F := ⟨λ X,
+begin
+  haveI : has_finite_limits (structured_arrow X F) :=
+    { out := λ J _ _, by { resetI, apply_instance } },
+  apply_instance
+end⟩
+
+namespace preserves_finite_limit_of_flat
+open structured_arrow
+open structured_arrow_cone
+variables {J : Type v₁} [small_category J] [fin_category J] {K : J ⥤ C}
+variables (F : C ⥤ D) [representably_flat F] {c : cone K} (hc : is_limit c) (s : cone (K ⋙ F))
+include hc
+
+/--
+(Implementation).
+Given a limit cone `c : cone K` and a cone `s : cone (K ⋙ F)` with `F` representably flat,
+`s` can factor through `F.map_cone c`.
+-/
+noncomputable def lift : s.X ⟶ F.obj c.X :=
+let s' := is_cofiltered.cone (to_diagram s ⋙ structured_arrow.pre _ K F) in
+s'.X.hom ≫ F.map (hc.lift ((cones.postcompose (eq_to_hom (by apply functor.hext; tidy))).obj
```

Comment:
Again I think using equality of functors is a bad idea here, especially because you're building an equality then converting it back to a morphism. If possible I think it would be a lot nicer to just make the nat trans you want and feed it into `cones.postcompose`, which might also make the later proofs (like the stuff in line 155) clearer.